### PR TITLE
JIT compiler, first cut

### PR DESCRIPTION
**Pull request by hrj**
:tada:

So, this is what I have been working on for the last few weeks!

Design wise, this is the first-cut, simple version. I have been testing/benchmarking/tweaking the implementation to see how far the design goes, so the implementation is quite stable.

The main purpose of the PR is to get some feedback from @jvilk, et al. Though, I believe it is stable enough to merge (after review).

## Design Overview

Currently, Doppio *interprets* JVM byte code one at a time. The opcodes are stored in an array and the interpreter walks the elements of the array sequentially. Each op-code is executed by invoking a corresponding function. The function is picked through a table indexed by the opcode.

One strategy for improving performance is to compile a sequence of opcodes into an equivalent JS function.

For example, consider this simple Java function:

```
public static int add(int a, int b, int c) {
  return a + b + c;
}
```

which after compilation produces this byte code:

```
public static int add(int, int, int);
  Code:
     0: iload_0
     1: iload_1
     2: iadd
     3: iload_2
     4: iadd
     5: ireturn
```

Instead of interpreting these opcodes one at a time, one could compile a JS function which does that:
```
function add() {
  iload_0();
  iload_1();
  iadd();
  iload_2();
  iadd();
  ireturn();
}
```

This avoids checking the opcodes of the method everytime. However, we can go further.  We can have stackless variants of the opcode implementations. For example, `stackless_iload_0` would return the loaded value rather than push it onto the stack. Similarly, `stackless_iadd` would accept two parameters and return the result, instead of using the stack. The stackless variant would look like:

```
function add() {
  const s0 = stackless_iload_0();
  const s1 = stackless_iload_1();
  const s2 = stackless_iadd(s0, s1);
  const s3 = stackless_iload_2();
  const s4 = stackless_iadd(s2, s3);
  stackless_ireturn(s4);
}
```

Further, the stackless functions could be inlined at the time of emitting, to give this form:

```
function add(frame, thread) {
  const s0 = frame.locals[0];
  const s1 = frame.locals[1];
  const s2 = s0 + s1;
  const s3 = frame.locals[2];
  const s4 = s2 + s3;
  thread.return(s4);
}
```

This form should improve performance since stack operations have been replaced with local variable assignments and the
Javascript VM can optimise this function fully.

Such a function can be emitted as a string and then compiled using JS `eval` or `new Function()`.

The above example is simplistic. There are some gotchas to consider:

*  If the function contains jumps then such a straightforward translation to JS is not possible.
*  If checks for yielding the thread are added, the continuation functions would create too many closures.

One solution is to create separate functions for each "trace", where a trace is defined as a contiguous block of opcodes that don't branch nor yield. A Java function might comprise of multiple such traces, and hence would have multiple JS functions associated with it.

## Tradeoffs

* Identifying traces, generating their JS functions and compiling those functions will increase the warm up time.
* Creating these JS functions would increase memory consumption.

To mitigate these factors, JITing can be restricted to frequently called functions. Currently a very simple heuristic is used: if the count of opcode executions in a method exceeds a threshold, then that method is targeted for JIT compilation.

### Symbolic stack
To keep track of symbols needed for stackless execution, a symbol stack is maintained similar to the operand stack. The difference is that the symbol stack tracks symbols at JIT time, while the operand stack tracks actual values at run-time.

When an opcode needs to pop two values from the operand stack, the JIT will instead pop two symbols from the symbol stack. When the symbol stack underflows, new symbols are created and initialised from the operand stack.

Similarly, when an opcode needs to push a value to the operand stack, the JIT emitted code will assign the value to a symbol which is then pushed on the symbol stack. This symbol can then be consumed by later opcodes in the `trace`. At the end of the trace, all symbols remaining on the symbol stack are mapped to pushes on the operand stack.

As an example, consider a trace that contains the following opcodes: `IADD`, `IMUL`. The first opcode needs two symbols from the symbolic stack. Since the symbolic stack is initially empty, we create two symbols, s1 and s2, push them to the symbolic stack, and emit code to initialise them:

```
var s1 = frame.opStack.pop();
var s2 = frame.opStack.pop();
```

These symbols are then available to the the `IADD` instruction, for which the following code can be emitted:

```
var s3 = s1 + s2;
```

Here, `s3` is a new symbol that represents the pushed value from `IADD`. This symbol will now be pushed on the symbol stack. When translating the next opcode `IMUL`, we again need two symbols to be popped from the symbol stack. But the stack contains only one symbol `s3`. Hence, a new symbol `s4` is allocated and pushed onto the stack, and code to initialise is emitted:

```
var s4 = frame.opStack.pop();
```

followed by the code for `IMUL`:

```
var s5 = s4 * s3;
```

Finally, at the end of the trace, all values represented by the symbols stack are pushed to the operand stack. In this case, only `s5` is present on the symbol stack, and thus:

```
frame.opStack.push(s5);
```

is emitted. Putting it all together, we get:

```
var s1 = frame.opStack.pop();
var s2 = frame.opStack.pop();
var s3 = s1 + s2;
var s4 = frame.opStack.pop();
var s5 = s4 * s3;
frame.opStack.push(s5);
```

Even in this small example we can see that one opStack push has been eliminated for the `IADD` instruction. The savings are larger for larger traces.

## Benchmarks

I have used the following benchmarks:

* Compilation of a simple Java program using `javac` in `Chromium` and `Firefox`
* Scimark 2 on `nodejs`
* Bubble sort and back-tracking N-Queens on `nodejs`
* Running a Scala program on nodejs

## Benchmark results

| javac   | Total time for JVM | Total time for JVM (after) |
| -----------   | ------------------ | -------------------------- |
| Chromium 49   | 31994              | 17534                      |
| Firefox 45.0.1| 38283              | 24582                      |

| Scimark2      | Score before      | Score after        |
| -----------   | ----------------  | ------------------ |
| node 5.7.1    | 2.194548131157776 | 23.46              |

| Nashorn       | Time before   | Time after     |
| -----------   | ------------  | -------------- |
| node 5.7.1    | 24.85 s       | 18.645s        |

| Abandon.jar  | Time before   | Time after     |
| ----------------- | ------------  | -------------- |
| node 5.7.1        | 90.53s        | 71.22s         |

| Bubble sort + nQueens | Time before   | Total time for JVM | Time after     | Total time for JVM (after) |
| --------------------- | ------------  | ------------------ | -------------- | -------------------------- |
| node 5.7.1            | 56011 ms      | 57.2 s             | 8248 ms        | 10.47 s                    |
### Future ideas

* Detection and optimisations of loops, if-else conditions, etc.
* Inlining functions calls.

### CODE REVIEW

**Comment by hrj** [comment]
The source code and binaries used for benchmarking, along with a README, are available [here](https://hrj.github.io/static/benchmark-files.tar.bz2).

File: src/jvm.ts [hunk 1]
```diff
@@ -427,6 +427,7 @@ class JVM {
         assert(false, `Invariant failure: Thread pool cannot be emptied post-JVM termination.`);
         return false;
       case JVMStatus.TERMINATING:
+        methods.dumpStats();
```
Comment (perimosocordiae) [inline src/jvm.ts]: This should probably be behind a flag, right?

File: src/threading.ts [hunk 2]
```diff
@@ -247,11 +270,16 @@ export class BytecodeStackFrame implements IStackFrame {
 
     // Run until we get the signal to return to the thread loop.
     while (!this.returnToThreadLoop) {
-      var op = code.readUInt8(this.pc);
-      if (!RELEASE && logging.log_level === logging.VTRACE) {
-        vtrace(`  ${this.pc} ${annotateOpcode(op, this, code, this.pc)}`);
+      // var op = code.readUInt8(this.pc);
+      var op = method.getOp(this.pc, code);
+      if (typeof op === 'function') {
+        op(this, thread, jitUtil);
```
Comment (perimosocordiae) [inline src/threading.ts]: Might be nice to have a VTRACE output that notes the call of a jitted function. IIRC, code behind RELEASE guards is stripped out for release builds, so it shouldn't cause any speed issues.

**Comment by jvilk** [comment]
My ability to review this is severely limited until after May 10th. I look forward to looking this over then!

**Comment by perimosocordiae** [comment]
Impressive work! It's a lot to review, so one thing that would help a lot is if you did some rebasing to structure the commits a little more nicely. Ideally, there should be:
 1. a commit that makes the non-JIT changes (like the new convenience functions for stack manipulation, and the signature change for the opcode annotation)
 2. a commit with the JIT, in a minimal working state
 3. commit(s) adding implementations of the various opcodes

It's a bit of a pain to shuffle commits around, I know, but it really does make a giant PR like this one easier to understand.

**Comment by hrj** [comment]
> It's a bit of a pain to shuffle commits around, I know, but it really does make a giant PR like this one easier to understand.

I agree, and thanks for looking!

@jvilk I am glad you mentioned a date.

I will work on a more structured PR and try to push it here before May 10th.

File: src/jvm.ts (hunk 1 above)
Comment (hrj) [inline src/jvm.ts]: Yup; I will wrap it in a flag. At the very minimum, it should have been under `!RELEASE`.

**Comment by hrj** [comment]
superseded by #443

**Pull request closed without merging**